### ISSUE TITLE
Fast native memory leak in WinHttpHandler on corefx

### ISSUE THREAD

**Issue by KKhurin** [dotnet/corefx#2762]
The leak is super-fast - 1GB+/minute and completely blocks WCF streaming scenarios.
Here are the top offending call stacks:
~~~
	ntdll!RtlAllocateHeap+14D (d:\blue\minkernel\ntos\rtl\heap.c, 1872)
	msvcrt!malloc+90 (d:\9147\minkernel\crts\crtw32\heap\malloc.c, 285)
	webio!zcalloc+20 (d:\9147\inetcore\lib\zlib12\zutil.c, 314)
	webio!updatewindow+39 (d:\9147\inetcore\lib\zlib12\inflate.c, 337)
	webio!inflate+12D7 (d:\9147\inetcore\lib\zlib12\inflate.c, 1137)
	webio!Decompress+56 (d:\9147\inetcore\lib\zlib12\iel1_zlib12_wrapper\zlib_wrapper.c, 218)
	webio!WapDecompProcessCompressedData+33 (d:\9147\net\webio\util\decompression.c, 510)
	webio!WaDecompDataNeeded+1C (d:\9147\net\webio\util\decompression.c, 580)
	webio!WapHandleParsedHttpResponse+1D74A (d:\9147\net\webio\http\recvresp.c, 1656)
	webio!WapHttpConnectionReceiveCompletion+258 (d:\9147\net\webio\http\httpconn.c, 2974)
	webio!WapTcpReceiveCompletionRoutine+89 (d:\9147\net\webio\conn\tcpconn.c, 3538)
	webio!WapTcpThreadPoolCompletionRoutine+3F (d:\9147\net\webio\conn\tcpconn.c, 822)
	webio!WapTpIoCompletionRoutine+B2 (d:\9147\net\webio\util\thrdpool.c, 370)
	KERNELBASE!BasepTpIoCallback+3A (d:\9147\minkernel\threadpool\kernel32\threadpool.c, 213)
	ntdll!TppIopExecuteCallback+D6 (d:\blue\minkernel\threadpool\ntdll\io.c, 371)
	ntdll!TppWorkerThread+52A (d:\blue\minkernel\threadpool\ntdll\worker.c, 1023)
	KERNEL32!BaseThreadInitThunk+24 (d:\9147\base\win32\client\thread.c, 78)
	ntdll!__RtlUserThreadStart+2F (d:\blue\minkernel\ntdll\rtlstrt.c, 1029)
	ntdll!_RtlUserThreadStart+1B (d:\blue\minkernel\ntdll\rtlstrt.c, 944)


---
	ntdll!RtlAllocateHeap+14D (d:\blue\minkernel\ntos\rtl\heap.c, 1872)
	webio!WaDecompCreateContext+AB (d:\9147\net\webio\util\decompression.c, 219)
	webio!WapHandleParsedHttpResponse+1D61A (d:\9147\net\webio\http\recvresp.c, 1496)
	webio!WapHttpConnectionReceiveCompletion+258 (d:\9147\net\webio\http\httpconn.c, 2974)
	webio!WapTcpReceiveCompletionRoutine+89 (d:\9147\net\webio\conn\tcpconn.c, 3538)
	webio!WapTcpThreadPoolCompletionRoutine+3F (d:\9147\net\webio\conn\tcpconn.c, 822)
	webio!WapTpIoCompletionRoutine+B2 (d:\9147\net\webio\util\thrdpool.c, 370)
	KERNELBASE!BasepTpIoCallback+3A (d:\9147\minkernel\threadpool\kernel32\threadpool.c, 213)
	ntdll!TppIopExecuteCallback+D6 (d:\blue\minkernel\threadpool\ntdll\io.c, 371)
	ntdll!TppWorkerThread+52A (d:\blue\minkernel\threadpool\ntdll\worker.c, 1023)
	KERNEL32!BaseThreadInitThunk+24 (d:\9147\base\win32\client\thread.c, 78)
	ntdll!__RtlUserThreadStart+2F (d:\blue\minkernel\ntdll\rtlstrt.c, 1029)
	ntdll!_RtlUserThreadStart+1B (d:\blue\minkernel\ntdll\rtlstrt.c, 944)
-----

	ntdll!RtlAllocateHeap+14D (d:\blue\minkernel\ntos\rtl\heap.c, 1872)
	msvcrt!malloc+90 (d:\9147\minkernel\crts\crtw32\heap\malloc.c, 285)
	webio!zcalloc+20 (d:\9147\inetcore\lib\zlib12\zutil.c, 314)
	webio!inflateInit2_+4F (d:\9147\inetcore\lib\zlib12\inflate.c, 164)
	webio!CreateDecompression+61 (d:\9147\inetcore\lib\zlib12\iel1_zlib12_wrapper\zlib_wrapper.c, 175)
	webio!WapHandleParsedHttpResponse+1D61A (d:\9147\net\webio\http\recvresp.c, 1496)
	webio!WapHttpConnectionReceiveCompletion+258 (d:\9147\net\webio\http\httpconn.c, 2974)
	webio!WapTcpReceiveCompletionRoutine+89 (d:\9147\net\webio\conn\tcpconn.c, 3538)
	webio!WapTcpThreadPoolCompletionRoutine+3F (d:\9147\net\webio\conn\tcpconn.c, 822)
	webio!WapTpIoCompletionRoutine+B2 (d:\9147\net\webio\util\thrdpool.c, 370)
	KERNELBASE!BasepTpIoCallback+3A (d:\9147\minkernel\threadpool\kernel32\threadpool.c, 213)
----

	ntdll!RtlAllocateHeap+14D (d:\blue\minkernel\ntos\rtl\heap.c, 1872)
	winhttp!WinHttpOpenRequestInternal+7C1 (d:\9147\net\winhttp\http\open.cxx, 717)
	winhttp!WinHttpOpenRequest+1F9 (d:\9147\net\winhttp\http\open.cxx, 881)
----

	ntdll!RtlAllocateHeap+14D (d:\blue\minkernel\ntos\rtl\heap.c, 1872)
	webio!WapCreateHttpRequest+C1 (d:\9147\net\webio\http\httpreq.c, 1139)
	webio!WebCreateHttpRequest+183 (d:\9147\net\webio\http\httpreq.c, 3192)
	winhttp!HTTP_USER_REQUEST::_SysSendRequest+D32 (d:\9147\net\winhttp\httpcore\usrreq.cpp, 2347)
	winhttp!HTTP_USER_REQUEST::_SendRequestWithDrainComplete+1A0 (d:\9147\net\winhttp\httpcore\usrreq.cpp, 1199)
	winhttp!HTTP_USER_REQUEST::SendRequest+3F3 (d:\9147\net\winhttp\httpcore\usrreq.cpp, 1529)
	winhttp!WinHttpSendRequest+4B9 (d:\9147\net\winhttp\api\sendapi.cxx, 265)

~~~

**Comment by davidsh** [comment]
@KKhurin Please provide a repro solution/code.  It's not really possible to diagnose this without a repro being posted.

**Comment by davidsh** [comment]
cc: @CIPop @SidharthNabar

**Comment by CIPop** [comment]
@KKhurin : I changed the title. PTAL - I'm not sure I'm right (Could be ClientWebSocket).

**Comment by davidsh** [comment]
This should be fixed now with #3096.

**Comment by davidsh** [comment]
@KKhurin The fix for the memory leaks is checked into the sources for System.Net.Http in the CoreFx repro.  

The fix is also available in the latest System.Net.Http/System.Net.Http.WinHttpHandler package version v4.0.1-beta-23310 from the CoreFx dev feed: https://www.myget.org/gallery/dotnet-core.
